### PR TITLE
fix: Notify response transformer fix

### LINKED ISSUE #1332: Make NotificationResponseTransformer respect atNotification.isEncrypted == false
### Describe the bug

If a client receives a notification where isEncrypted is explicitly false, it nonetheless continues to treat it as though isEncrypted were true

### Steps to reproduce

Came across this bug when I was running an sshnpd with legacy atKeys at the same time as I was doing enroll requests; the enroll request notifications explicitly have isEncrypted == false (which is correct) but NotificationResponseTransformer nonetheless tries to decrypt the data and of course fails

### Expected behavior

explicit isEncrypted == false should be respected

### Additional context

See also https://github.com/atsign-foundation/at_server/issues/1944

### PR DESCRIPTION
**Pull request by murali-shris**
fixes: https://github.com/atsign-foundation/at_client_sdk/issues/1332
**- What I did**
- added check in NotificationResponseTransformer to decrypt the notification value only if isEncrypted is NOT false.
**- How I did it**
- added a check before decrypting notification value. decrypt notification value only if isEncrypted is null or true.
- added/modified unit tests in notification_service_test.dart
**- How to verify it**
- tests in notification_service_test.dart should pass

### CODE REVIEW

File: packages/at_client/lib/src/transformer/response_transformer/notification_response_transformer.dart [hunk 1]
```diff
@@ -76,6 +77,12 @@ class NotificationResponseTransformer
     return atNotification;
   }
 
+  // Decrypt the notification key only if isEncrypted flag is set to true or
+  // null (for backward compatibility - default behavior is decrypt the notification key)
+  bool _shouldDecryptKey(bool? isEncrypted) {
```
Comment (VJag) [inline packages/at_client/lib/src/transformer/response_transformer/notification_response_transformer.dart]: I think the doc should be like "Returns true when a notification is encrypted and needs a decryption etc.." also prefer this method taking a notification as an argument

Comment (VJag) [reply packages/at_client/lib/src/transformer/response_transformer/notification_response_transformer.dart]: Then we can write tests around that flag being null, true and false

Comment (murali-shris) [reply packages/at_client/lib/src/transformer/response_transformer/notification_response_transformer.dart]: > I think the doc should be like "Returns true when a notification is encrypted and needs a decryption etc.." also prefer this method taking a notification as an argument

there are two fields on which we do encryption/decryption. 1) notification key 2) notification value. hence i didn't add a method to decrypt by passing in notification object

Comment (VJag) [reply packages/at_client/lib/src/transformer/response_transformer/notification_response_transformer.dart]: Not to decrypt, it is the bool method.

**Review by gkc — CHANGES_REQUESTED** [review]
The changes so far don't address the primary problem as described in the associated issue, which is that the code will attempt to decrypt a standard `update` notification even if isEncrypted is false

```
if ((atNotification.value.isNotNull) &&
        (config.shouldDecrypt && atNotification.id != '-1') &&
        // The shared_key (which is a reserved key) has different decryption process
        // and is not a user created key.
        // Hence do not decrypt if key's are reserved keys
        AtKey.getKeyType(atKey.toString()) != KeyType.reservedKey) {
      // decrypt the value
      atNotification.value =
          await _getDecryptedValue(atKey, atNotification.value!);
      return atNotification;
    }
```

Please add unit tests for standard update notifications to verify (1) that decryption will be attempted if isEncrypted is null (2) that decryption will be attempted if isEncrypted is true (3) that decryption will not be attempted if isEncrypted is false and make implementation changes required to make test (3) pass

**Comment by murali-shris** [comment]
> The changes so far don't address the primary problem as described in the associated issue, which is that the code will attempt to decrypt a standard `update` notification even if isEncrypted is false
> 
> ```
> if ((atNotification.value.isNotNull) &&
>         (config.shouldDecrypt && atNotification.id != '-1') &&
>         // The shared_key (which is a reserved key) has different decryption process
>         // and is not a user created key.
>         // Hence do not decrypt if key's are reserved keys
>         AtKey.getKeyType(atKey.toString()) != KeyType.reservedKey) {
>       // decrypt the value
>       atNotification.value =
>           await _getDecryptedValue(atKey, atNotification.value!);
>       return atNotification;
>     }
> ```
> 
> Please add unit tests for standard update notifications to verify (1) that decryption will be attempted if isEncrypted is null (2) that decryption will be attempted if isEncrypted is true (3) that decryption will not be attempted if isEncrypted is false and make implementation changes required to make test (3) pass

@gkc in current trunk code, there are two flags used in notification response transformer, AtNotification --> isEncrypted and NotificationConfig --> shouldDecrypt.
isEncrypted is used for notification text where the value is sent as part of notification key. 
shouldDecrypt is used for standard update notification. 
Should I revert changes I did for isEncrypted and make the changes for  shouldDecrypt flag or use isEncrypted for both scenarios?

**Comment by gkc** [comment]
@murali-shris
here's the relevant block of code that needs to be changed:
```
else if ((atNotification.value.isNotNull) &&
        (config.shouldDecrypt && atNotification.id != '-1') &&
        // The shared_key (which is a reserved key) has different decryption process
        // and is not a user created key.
        // Hence do not decrypt if key's are reserved keys
        AtKey.getKeyType(atKey.toString()) != KeyType.reservedKey) {
      // decrypt the value
      atNotification.value =
          await _getDecryptedValue(atKey, atNotification.value!);
      return atNotification;
    }
```

- isEncrypted is part of the notification message's metadata
- currently, isEncrypted is **not checked** - i.e. the notification's value is always decrypted 
- the change required is to check if isEncrypted == false, and do not decrypt in that case, regardless of what shouldDecrypt says, since there is no point in even trying to decrypt a message which is not encrypted

**Comment by gkc** [comment]
i.e. change this code from the above block from
```
      // decrypt the value
      atNotification.value =
          await _getDecryptedValue(atKey, atNotification.value!);
```

to

```
      if (atNotification.isEncrypted != false) {
        // decrypt the value
        atNotification.value =
        await _getDecryptedValue(atKey, atNotification.value!);
      }
```

File: packages/at_client/lib/src/transformer/response_transformer/notification_response_transformer.dart [hunk 2]
```diff
@@ -62,15 +61,18 @@ class NotificationResponseTransformer
...
+        atNotification.id != '-1' &&
         // The shared_key (which is a reserved key) has different decryption process
         // and is not a user created key.
         // Hence do not decrypt if key's are reserved keys
         AtKey.getKeyType(atKey.toString()) != KeyType.reservedKey) {
-      // decrypt the value
-      atNotification.value =
-          await _getDecryptedValue(atKey, atNotification.value!);
+      // decrypt the notification value only if isEncrypted is not set to false
+      if (atNotification.isEncrypted != false) {
```
Comment (gkc) [inline packages/at_client/lib/src/transformer/response_transformer/notification_response_transformer.dart]: If it's not false then we need to decrypt - but only if `config.shouldDecrypt` is true i.e. this line should read, I think, `if (atNotification.isEncrypted != false && config.shouldDecrypt)`

File: packages/at_client/test/notification_service_test.dart [hunk 3]
```diff
@@ -276,9 +276,35 @@ void main() {
...
+          await notificationResponseTransformer.transform(Tuple()
+            ..one = atNotification
+            ..two = (NotificationConfig()
+              ..regex = '.*'
+              ..shouldDecrypt = true));
+      expect(transformedNotification.value, 'decryptedValue');
+    });
+
+    test(
+        'A test to verify notification value is decrypted when isEncrypted is set to true',
```
Comment (gkc) [inline packages/at_client/test/notification_service_test.dart:305]: We need another test, to verify that the notification value is not decrypted when (a) isEncrypted is true, and (b) shouldDecrypt is false

**Review by gkc — CHANGES_REQUESTED** [review]

**Review by gkc — APPROVED** [review]

**Pull request merged**